core/dbt/parser/manifest.py

```diff
@@ -367,82 +318,132 @@ def read_parse_results(self) -> Optional[ParseResult]:
-        # list is created
+        process_sources(self.manifest, project_name)
+        process_refs(self.manifest, project_name)
+        process_docs(self.manifest, self.root_project)
+
+    def update_manifest(self) -> Manifest:
         start_patch = time.perf_counter()
-        sources = patch_sources(self.results, self.root_project)
+        # patch_sources converts the UnparsedSourceDefinitions in the
+        # Manifest.sources to ParsedSourceDefinition via 'patch_source'
+        # in SourcePatcher
+        sources = patch_sources(self.root_project, self.manifest)
+        self.manifest.sources = sources
+        # ParseResults had a 'disabled' attribute which was a dictionary
+        # which is now named '_disabled'. This used to copy from
+        # ParseResults to the Manifest. Can this be normalized so
+        # there's only one disabled?
+        disabled = []
+        for value in self.manifest._disabled.values():
+            disabled.extend(value)
+        self.manifest.disabled = disabled
         self._perf_info.patch_sources_elapsed = (
             time.perf_counter() - start_patch
         )
-        disabled = []
-        for value in self.results.disabled.values():
-            disabled.extend(value)
 
-        nodes: MutableMapping[str, ManifestNode] = {
-            k: v for k, v in self.results.nodes.items()
-        }
+        self.manifest.selectors = self.root_project.manifest_selectors
 
-        manifest = Manifest(
-            nodes=nodes,
-            sources=sources,
-            macros=self.results.macros,
-            docs=self.results.docs,
-            exposures=self.results.exposures,
-            metadata=self.root_project.get_metadata(),
-            disabled=disabled,
-            files=self.results.files,
-            selectors=self.root_project.manifest_selectors,
-        )
-        manifest.patch_nodes(self.results.patches)
-        manifest.patch_macros(self.results.macro_patches)
+        # do the node and macro patches
+        self.manifest.patch_nodes()
+        self.manifest.patch_macros()
+
+        # process_manifest updates the refs, sources, and docs
         start_process = time.perf_counter()
-        self.process_manifest(manifest)
+        self.process_manifest()
 
         self._perf_info.process_manifest_elapsed = (
             time.perf_counter() - start_process
         )
 
-        return manifest
+        return self.manifest
+
+    # TODO: this should be calculated per-file based on the vars() calls made in
+    # parsing, so changing one var doesn't invalidate everything. also there should
+    # be something like that for env_var - currently changing env_vars in way that
+    # impact graph selection or configs will result in weird test failures.
+    # finally, we should hash the actual profile used, not just root project +
+    # profiles.yml + relevant args. While sufficient, it is definitely overkill.
```

Comment:
This is a case where we're being overzealous, and fully re-parsing the project when we could make partial parsing smarter. Therefore, this is in the "slow + correct" bucket, rather than the "fast + incorrect" bucket. We shouldn't lose track of this, but for now, we should be prioritizing edge cases in the latter category.